lib/rom/setup/auto_registration.rb

```diff
@@ -44,21 +44,67 @@ def mappers
     def load_entities(entity)
       Dir[globs[entity]].map do |file|
         require file
-        Inflector.constantize(
-          const_name(component_dirs.fetch(entity), file)
+        klass_name = case
+        when namespace.class == String
+          CustomNamespaceStrategy.new(namespace: namespace, file: file).call
+        when namespace == true
+          WithNamespaceStrategy.new(file: file, directory: directory).call
+        when namespace == false
+          NoNamespaceStrategy.new(file: file, directory: directory, entity: component_dirs.fetch(entity)).call
+        end
+        Inflector.constantize(klass_name)
+      end
+    end
+
+    class CustomNamespaceStrategy
+      include Options
+      option :file, reader: true, type: String
+      option :namespace, reader: true, type: String
+
+      def call
+        "#{namespace}::#{Inflector.camelize(filename).sub(EXTENSION_REGEX, '')}"
+      end
+
+      private
+
+      attr_reader :namespace, :file
+
+      def filename
+        Pathname.new(file).basename.to_s
+      end
+    end
+
+    class WithNamespaceStrategy
+      include Options
+      option :directory, reader: true, type: Pathname
+      option :file, reader: true, type: String
+
+      def call
+        Inflector.camelize(
+          file.sub(/^#{directory.dirname}\//, '').sub(EXTENSION_REGEX, '')
         )
       end
+
+      private
+
+      attr_reader :directory, :file
     end
 
-    def const_name(entity, file)
-      name =
-        if namespace
-          file.sub(/^#{directory.dirname}\//, '')
-        else
-          file.sub(/^#{directory}\/#{entity}\//, '')
-        end.sub(EXTENSION_REGEX, '')
+    class NoNamespaceStrategy
+      include Options
+      option :directory, reader: true, type: Pathname
+      option :file, reader: true, type: String
+      option :entity, reader: true, type: Symbol
+
+      def call
+        Inflector.camelize(
+          file.sub(/^#{directory}\/#{entity}\//, '').sub(EXTENSION_REGEX, '')
+        )
+      end
+
+      private
```

Comment:
Seems like we could have an abstract strategy to reduce some level of duplication ie:

``` ruby
class Strategy
  include Options
  option :file, reader: true, type: String
end
```